File: src/FSOps.Core/Entities/Pilot.cs
namespace FSOps.Core.Entities;

public class Pilot
{
    public Guid Id { get; set; }

    public Guid AirlineId { get; set; }

    public string Name { get; set; } = string.Empty;

    public bool IsPlayer { get; set; }

    public decimal MonthlySalary { get; set; }

    public double HoursFlown { get; set; }

    public double SkillRating { get; set; } = 50;

    /// <summary>
    /// When this pilot last completed a flight (player or virtual) - the ONLY clock
    /// <see cref="FSOps.Core.Scheduling.PilotSkillCalculator"/>'s idle decay is ever measured
    /// against, deliberately never wall-clock time since the app was last opened. A pilot on a
    /// standing schedule keeps this fresh while the app is closed (the wall-clock catch-up model
    /// resolves their flights regardless), so decay only ever reaches a pilot deliberately left
    /// with no schedule - see docs/PLAN.md "Progression - reputation and pilot skill". Null for a
    /// pilot who has never flown a single sector.
    /// </summary>
    public DateTimeOffset? LastFlewUtc { get; set; }

    public PilotStatus Status { get; set; } = PilotStatus.Available;

    public DateTimeOffset CreatedUtc { get; set; }

    public DateTimeOffset? DeletedUtc { get; set; }
}

File: src/fsops-web/src/types/pilot.ts
export type PilotStatus = 'Available' | 'Flying' | 'Inactive'

/**
 * A saved weekly pattern that has quietly stopped producing flights, because its aircraft is parked
 * at an airport no leg in the pattern departs from - so nothing in the schedule can ever move it
 * again. Not an error and not a validation failure: the schedule is still valid and still
 * repeating, and it resumes on its own the moment the aircraft is back. See ScheduleStallDetector
 * for why this is narrower than "the aircraft isn't where the week starts" (a pattern that visits
 * wherever the aircraft is standing repairs itself, and warning about it would be crying wolf).
 */
export interface ScheduleStall {
  fleetAircraftId: string
  registration: string
  /** Where the aircraft actually is. */
  locationIcao: string
  /** Where the earliest leg of the week departs from. */
  patternStartIcao: string
  /** Server-composed sentence: what is wrong, and the two ways out. */
  message: string
}

/**
 * One virtual pilot, as returned by GET /pilots. The trailing three fields are weekly
 * projections computed from the pilot's saved schedule - the backend may not populate them
 * yet ("may be null in the first cut"), so callers must treat their absence as "not known"
 * rather than zero.
 */
export interface PilotSummary {
  id: string
  name: string
  isPlayer: boolean
  monthlySalary: number
  hoursFlown: number
  /** Recomputed live on every fetch (see PilotEndpoints.ListAsync) - never stale between the
   *  resolver's periodic decay passes, so this is always the true-right-now figure. */
  skillRating: number
  /** What hoursFlown alone would give this pilot with no idle decay applied - lets the UI show
   *  "earned X, currently Y" when decay has pulled skillRating below it. */
  earnedSkillRating: number
  /** Null for the player pilot (their own record never decays) and for a virtual pilot who has
   *  never flown yet. */
  lastFlewUtc: string | null
  /** Days since lastFlewUtc, null on the same conditions as lastFlewUtc. */
  idleDays: number | null
  /** True once idle time has passed the grace period and skillRating is actively decaying. */
  isDecaying: boolean
  /** Days of grace left before decay starts, 0 once isDecaying is true. Null on the same
   *  conditions as lastFlewUtc. */
  decayGraceDaysRemaining: number | null
  /** Derived server-side on every fetch from this pilot's in-progress flight and schedule, never
   *  read from a stored column - see PilotStatusCalculator. */
  status: PilotStatus
  createdUtc: string
  sectorsPerWeek?: number | null
  weeklyEstimatedRevenue?: number | null
  weeklyEstimatedCost?: number | null
  /** Empty for the overwhelmingly common case. One entry per aircraft on this pilot's schedule that
   *  has stalled - see ScheduleStall. */
  scheduleStalls?: ScheduleStall[]
}

/** POST /pilots response: the new pilot plus the airline's cash balance after the hiring cost. */
export interface HirePilotResponse {
  pilot: PilotSummary
  cashBalance: number
}
